server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java

```diff
@@ -20,93 +20,113 @@
+  /**
+   * Duplicated blocks in the current file are either {@link Duplication#getOriginal()} or {@link Duplication#getDuplicates()}
+   * when the {@link org.sonar.server.computation.duplication.Duplicate} is a {@link InnerDuplicate}.
+   * <p>
+   * The returned list is mutable on purpose because it will be sorted.
+   * </p>
+   *
+   * @see {@link #createIndexOfDuplicatedTextBlocks(Collection)}
+   */
+  private static List<TextBlock> extractAllDuplicatedTextBlocks(Collection<Duplication> duplications) {
+    List<TextBlock> duplicatedBlock = new ArrayList<>(duplications.size());
+    for (Duplication duplication : duplications) {
+      duplicatedBlock.add(duplication.getOriginal());
+      for (InnerDuplicate duplicate : from(duplication.getDuplicates()).filter(InnerDuplicate.class)) {
+        duplicatedBlock.add(duplicate.getTextBlock());
+      }
+    }
+    return duplicatedBlock;
   }
 
-  private static int length(BatchReport.TextRange range) {
-    return (range.getEndLine() - range.getStartLine()) + 1;
-  }
+  private static class TextBlockContainsLine implements Predicate<Map.Entry<TextBlock, Integer>> {
+    private final int line;
 
-  private static Map<BatchReport.TextRange, Integer> createDuplicationIdsByRange(List<BatchReport.Duplication> duplications) {
-    Map<BatchReport.TextRange, Integer> map = newHashMap();
-    int blockId = 1;
-    for (BatchReport.Duplication duplication : duplications) {
-      map.put(duplication.getOriginPosition(), blockId);
-      blockId++;
-      for (BatchReport.Duplicate duplicate : duplication.getDuplicateList()) {
-        if (isDuplicationOnSameFile(duplicate)) {
-          map.put(duplicate.getRange(), blockId);
-          blockId++;
-        }
-      }
+    public TextBlockContainsLine(int line) {
+      this.line = line;
+    }
+
+    @Override
+    public boolean apply(@Nonnull Map.Entry<TextBlock, Integer> input) {
+      return isLineInBlock(input.getKey(), line);
     }
-    return map;
   }
 
-  private static class DuplicationComparator implements Comparator<BatchReport.Duplication>, Serializable {
+  private enum MapEntryToBlockId implements Function<Map.Entry<TextBlock, Integer>, Integer> {
+    INSTANCE;
+
     @Override
-    public int compare(BatchReport.Duplication d1, BatchReport.Duplication d2) {
-      if (d1.getOriginPosition().getStartLine() == d2.getOriginPosition().getStartLine()) {
-        return Integer.compare(length(d1.getOriginPosition()), length(d2.getOriginPosition()));
-      } else {
-        return Integer.compare(d1.getOriginPosition().getStartLine(), d2.getOriginPosition().getStartLine());
-      }
+    @Nonnull
+    public Integer apply(@Nonnull Map.Entry<TextBlock, Integer> input) {
+      return input.getValue();
     }
   }
 
+  private static class TextBlockIndexGenerator implements Function<TextBlock, Integer> {
+    int i = 1;
+
+    @Nullable
+    @Override
+    public Integer apply(TextBlock input) {
+      return i++;
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this increment or correct the code not to waste it. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2123)